### PR TITLE
Add GetWorkflowHistoryReverse API

### PR DESCRIPTION
**Pull request by Ardagan**
**What changed?**
Add fields to support retrieving workflow history in reverse order.

**Why?**
More efficient way to query large histories.

**How did you test it?**
Manual, Automated tests

**Potential risks**
N/A

### CODE REVIEW

File: temporal/api/history/v1/message.proto [hunk 1]
```diff
@@ -642,4 +642,5 @@ message HistoryEvent {
 
 message History {
     repeated HistoryEvent events = 1;
+    int64 transaction_id = 2;
```
Comment (yiminc) [inline temporal/api/history/v1/message.proto:645]: I don't think we need this to be exposed in public api.

File: temporal/api/workflowservice/v1/request_response.proto [hunk 2]
```diff
@@ -191,6 +191,23 @@ message GetWorkflowExecutionHistoryResponse {
...
 }
 
+message GetWorkflowExecutionHistoryReverseRequest {
+    string namespace = 1;
+    temporal.api.common.v1.WorkflowExecution execution = 2;
+    int32 maximum_page_size = 3;
+    // If a `GetWorkflowExecutionHistoryResponse` or a `PollWorkflowTaskQueueResponse` had one of
+    // these, it should be passed here to fetch the next page.
+    bytes next_page_token = 4;
+    bool skip_archival = 7;
```
Comment (cretz) [inline temporal/api/workflowservice/v1/request_response.proto:201]: Intentionally keeping same index for `skip_archival` as `GetWorkflowExecutionHistoryRequest`?

File: temporal/api/workflowservice/v1/service.proto [hunk 3]
```diff
@@ -94,6 +94,12 @@ service WorkflowService {
     // `NotFound` if the specified workflow execution is unknown to the service.
     rpc GetWorkflowExecutionHistory (GetWorkflowExecutionHistoryRequest) returns (GetWorkflowExecutionHistoryResponse) {
     }
+    
+    // GetWorkflowExecutionHistoryReverse returns the history of specified workflow execution in reverse 
+    // order (starting from last event). Fails with`NotFound` if the specified workflow execution is 
+    // unknown to the service.
+    rpc GetWorkflowExecutionHistoryReverse (GetWorkflowExecutionHistoryReverseRequest) returns (GetWorkflowExecutionHistoryReverseResponse) {
```
Comment (cretz) [inline temporal/api/workflowservice/v1/service.proto:101]: Is this not clearer as just a boolean on `GetWorkflowExecutionHistoryRequest` (and each successive page request after)? Could even embed the direction in the token (even if just for verifying the bool is the same/expected on each page call, not necessarily for affecting the direction).

File: temporal/api/workflowservice/v1/request_response.proto (hunk 2 above)
Comment (Ardagan) [reply temporal/api/workflowservice/v1/request_response.proto:201]: Yes. Keeping IDs compatible will not make it worse.

File: temporal/api/workflowservice/v1/service.proto (hunk 3 above)
Comment (Ardagan) [reply temporal/api/workflowservice/v1/service.proto:101]: Getting history in reverse has less functionality, performance and is effectively a separate code path at this point. It is cleaner to have a separate api.

Comment (yiminc) [reply temporal/api/workflowservice/v1/service.proto:101]: cc @stevekinney

File: temporal/api/workflowservice/v1/request_response.proto (hunk 2 above)
Comment (Ardagan) [reply temporal/api/workflowservice/v1/request_response.proto]: you are correct.

**Review by yycptt — APPROVED** [review]

**Pull request merged**